### PR TITLE
cp2k: add a separate MakefileBuilder

### PR DESCRIPTION
**Pull request by alalazo**
follow up from #42009

Modifications:
- [x] Extracted a separate `MakefileBuilder`
- [x] Removed deprecated versions, cleaned-up the recipe
- [x] Factored together the `GPU_MAP` dictionary, which is now used by both builders

### CODE REVIEW

**Comment by alalazo** [comment]
@lorisercole @mtaillefumier Here's the clean-up PR discussed in #42009 I just tried two simple builds (one with `cmake` and one with `make`), so please test this and let me know if there's anything that needs fixing or if some of the changes are unwanted :slightly_smiling_face:

**Review by mtaillefumier — COMMENTED** [review]
thanks. So far i can not find any obvious issue with these changes. I will test both builds and report back here. I leave the PR in its current status and will change it once I tested few builds combinations

**Comment by lorisercole** [comment]
In [line 382](https://github.com/alalazo/spack/blob/faaf09181423eff2fb60ee22365c1fbe36cefd0e/var/spack/repos/builtin/packages/cp2k/package.py#L382) I would change this assignment:
```python
        cppflags = ["-D__FFTW3", "-I{0}".format(fftw_header_dir)]
```
to
```python
        if spec["fftw-api"].name in ("intel-mkl", "intel-parallel-studio", "intel-oneapi-mkl"):
            cppflags = ["-D__FFTW3_MKL", "-I{0}".format(fftw_header_dir)]
        else:
            cppflags = ["-D__FFTW3", "-I{0}".format(fftw_header_dir)]
```
since the [CP2K install guide](https://github.com/cp2k/cp2k/blob/master/INSTALL.md#2f-fftw-optional-improved-performance-of-ffts) suggests to use that value when using MKL libraries.

**Comment by lorisercole** [comment]
I would change [Line 501](https://github.com/alalazo/spack/blob/faaf09181423eff2fb60ee22365c1fbe36cefd0e/var/spack/repos/builtin/packages/cp2k/package.py#L501) to this, to support `hpcx-mpi` (which is based on OpenMPI):
```python
                mpi_impl = "openmpi" if spec["mpi"].name in ["openmpi", "hpcx-mpi"] else "intelmpi"
```

**Comment by lorisercole** [comment]
For the CMake builder recipe, maybe we can add the build type (I'm not sure it is automatic)?
```python
            self.define_from_variant("CMAKE_BUILD_TYPE", "build_type"),
```

**Comment by mtaillefumier** [comment]
> For the CMake builder recipe, maybe we can add the build type (I'm not sure it is automatic)?
> 
> ```python
>             self.define_from_variant("CMAKE_BUILD_TYPE", "build_type"),
> ```

I agree. I forgot to add it in the recipe.

**Comment by alalazo** [comment]
> For the CMake builder recipe, maybe we can add the build type [ ... ]

That is not needed:

https://github.com/spack/spack/blob/75e96b856eba55f7a47a2f446b3cd4e8f3f90e3c/lib/spack/spack/build_systems/cmake.py#L272-L277

**Comment by mtaillefumier** [comment]
I get this error 

```
==> Error: AttributeError: 'str' object has no attribute '_cmp_iter'

The 'cp2k' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure

/home/mathieut/Projects/codes/spack/var/spack/repos/builtin/packages/cp2k/package.py:501, in edit:
        498            # while intel-mkl has a mpi variant and adds the scalapack
        499            # libs to its libs, intel-oneapi-mkl does not.
        500            if spec["scalapack"].name == "intel-oneapi-mkl":
  >>    501                mpi_impl = "openmpi" if spec["mpi"] == "openmpi" else "intelmpi"
        502                scalapack = [
        503                    join_path(
        504                        spec["intel-oneapi-mkl"].libs.directories[0], "libmkl_scalapack_lp64.so"
```
when I execute 

`spack install cp2k@2023.2 build_system=makefile +libxc +libint +mpi ^intel-oneapi-mkl+cluster ^openmpi`

**Comment by lorisercole** [comment]
> I get this error
> 
> ```
> ==> Error: AttributeError: 'str' object has no attribute '_cmp_iter'
> 
> The 'cp2k' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure
> 
> /home/mathieut/Projects/codes/spack/var/spack/repos/builtin/packages/cp2k/package.py:501, in edit:
>         498            # while intel-mkl has a mpi variant and adds the scalapack
>         499            # libs to its libs, intel-oneapi-mkl does not.
>         500            if spec["scalapack"].name == "intel-oneapi-mkl":
>   >>    501                mpi_impl = "openmpi" if spec["mpi"] == "openmpi" else "intelmpi"
>         502                scalapack = [
>         503                    join_path(
>         504                        spec["intel-oneapi-mkl"].libs.directories[0], "libmkl_scalapack_lp64.so"
> ```
> 
> when I execute
> 
> `spack install cp2k@2023.2 build_system=makefile +libxc +libint +mpi ^intel-oneapi-mkl+cluster ^openmpi`

Yes, this is a bug. See my suggestion above, it should be `spec["mpi"].name`

**Comment by mtaillefumier** [comment]
i've another one... 
`
spack install cp2k@2023.2 +libxc +libint +mpi ^intel-oneapi-mkl+cluster ^openmpi`

does not work either. cmake fails but I do not know yet if it coming from my install or not. this one works

`
spack install cp2k@master +libxc +libint +mpi ^intel-oneapi-mkl+cluster ^openmpi`

**Comment by alalazo** [comment]
@lorisercole @mtaillefumier I'll fix those. Note though that they are on lines not changed by the PR, so they were in `develop` already :wink:

**Comment by mtaillefumier** [comment]
> @lorisercole @mtaillefumier I'll fix those. Note though that they are on lines not changed by the PR, so they were in `develop` already 😉

that's alright. i never had this issue because I always use cmake to build cp2k. Surprizingly no-one found out about it either.

**Comment by alalazo** [comment]
With respect to https://github.com/spack/spack/pull/42130#issuecomment-1900634356 This one fails on:

Screenshot from 2024-01-19 18-28-08

Apparently it searches for `fftw`, and can't find it. `fftw` is not in the graph and the virtual is provided by `intel-oneapi-mkl`.

**Comment by alalazo** [comment]
The issue for me is the setup. Without further constraints, I end up with:
- openblas: for blas and lapack
- openmpi: for mpi
- intel-oneapi-mkl: for fftw-api

That doesn't seem to configure correctly for me, with the current CMake recipe. If I use:
```yaml
spack:
  # add package specs to the `specs` list
  specs:
  - cp2k@2023.2+libint+libxc+mpi ^intel-oneapi-mkl+cluster ^openmpi
  view: true
  concretizer:
    unify: true
  packages:
    blas:
      require: intel-oneapi-mkl
    lapack:
      require: intel-oneapi-mkl
    fftw-api:
      require: intel-oneapi-mkl
```
I can build correctly.

**Comment by mtaillefumier** [comment]
that's really strange.

**Comment by mtaillefumier** [comment]
```
lapack = spec["lapack"]
        blas = spec["blas"]

        if blas.name in ["intel-mkl", "intel-parallel-studio", "intel-oneapi-mkl"]:
            args += ["-DCP2K_BLAS_VENDOR=MKL"]
            if sys.platform == "darwin":
                args += [
                    self.define("CP2K_BLAS_VENDOR", "CUSTOM"),
                    self.define("CP2K_SCALAPACK_VENDOR", "GENERIC"),
                    self.define(
                        "CP2K_SCALAPACK_LINK_LIBRARIES", spec["scalapack"].libs.joined(";")
                    ),
                ]
            else:
                args += ["-DCP2K_SCALAPACK_VENDOR=MKL"]
```

this should ensure that the blas and lapack are set properly.

**Comment by mtaillefumier** [comment]
and the only to trigger the generic case is on macs.  I am really puzzled by this. Then there is this problem fftw that I thought i had fixed but apparently not.

**Comment by alalazo** [comment]
Blas and lapack are set correctly, but then we hit https://github.com/cp2k/cp2k/blob/bfded9abfe01386511154a41a7615142aecd0120/CMakeLists.txt#L637-L639

It seems that using `openblas` for `lapack` and `intel-oneapi-mkl` for `fftw` is not working. Honestly not sure if that configuration even make sense in reality, but it's one that makes sense for the solver, given the constraint we gave it :slightly_smiling_face:

**Comment by lorisercole** [comment]
If I build with Makefile
```
cp2k@2023.2%gcc +mpi +openmp +elpa +libint +libxc +spglib smm=libxsmm build_system=makefile ^openmpi ^fftw@3 ^netlib-scalapack ^openblas
```
I hit this linking error due to `libxsmm`:
```bash
     817    /usr/bin/mpif90 -c -O2 -funroll-loops -ftree-vectorize -ffree-form -ffree-line-length-none -ggdb -fopenmp -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libint-2.6.0-gzzjdrt4h4fqli2hi75vs34gsxy6lh32/include -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libint-2.6.0-gzzjdrt4h4fqli2hi75vs34gsxy6lh32/include/
            libint2 -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/boost-1.84.0-hwwnraam3zmigqp2yqy4dipb7cz7ib4q/include -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libxc-6.2.2-4s2i4pso4wjmk3tfpisxi4hqg2ybdta2/include -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/elpa-2023.05.001-kadh2
            ux2h6p67dkebz6dwzwaiysvqupi/include/elpa_openmp-2023.05.001/modules -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/elpa-2023.05.001-kadh2ux2h6p67dkebz6dwzwaiysvqupi/include/elpa_openmp-2023.05.001/elpa -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libxsmm-1.17-wzwhf5uriw2g4zopxntmjlbx7wx4pk7n/include -D__F
            FTW3 -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/fftw-3.3.10-rvdcxs5iywzcwzrxzwzamexdke5czoki/include -D__parallel -D__SCALAPACK -D__LIBINT -D__LIBXC -D__ELPA=202305 -D__LIBXSMM -D__SPGLIB -D__COMPILE_ARCH="\"linux-ubuntu22.04-skylake-gcc\"" -D__COMPILE_DATE="\"Fri Jan 19 16:33:07 CET 2024\"" -D__COMPILE_HOST="\"DESKTOP-HSCRDM6\"" -D
            __COMPILE_REVISION="\"git:65cca6c\"" -D__DATA_DIR="\"/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/cp2k-2023.2-3br2elq4zz6jwu7apjgupo7vab5tfhmm/share/data\"" -D__SHORT_FILE__="\"distribution_2d_types.F\"" -I'/tmp/lercole/spack-stage/spack-stage-cp2k-2023.2-3br2elq4zz6jwu7apjgupo7vab5tfhmm/spack-src/src/' -I'/tmp/lercole/spack-stage/spack
            -stage-cp2k-2023.2-3br2elq4zz6jwu7apjgupo7vab5tfhmm/spack-src/obj/linux-ubuntu22.04-skylake-gcc/psmp/exts/dbcsr' distribution_2d_types.F90
  >> 818    /usr/bin/ld: /home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libxsmm-1.17-wzwhf5uriw2g4zopxntmjlbx7wx4pk7n/lib/libxsmm.so: undefined reference to `dgemm_'
  >> 819    /usr/bin/ld: /home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libxsmm-1.17-wzwhf5uriw2g4zopxntmjlbx7wx4pk7n/lib/libxsmm.so: undefined reference to `sgemm_'
```

**Comment by mtaillefumier** [comment]
> Blas and lapack are set correctly, but then we hit https://github.com/cp2k/cp2k/blob/bfded9abfe01386511154a41a7615142aecd0120/CMakeLists.txt#L637-L639
> 
> It seems that using `openblas` for `lapack` and `intel-mkl` for `fftw` is not working. Honestly not sure if that configuration even make sense in reality, but it's one that makes sense for the solver, given the constraint we gave it 🙂

this configuration does not make sense. I either use `mkl + fftw` or `mkl + mkl+fftw` but `openblas+mkl-fftw` is dangerous in many ways as mkl-fftw is integral part `libmkl[*].,so`. To be clearer we do not know which implementation of blas we  use when this configuration `openblas+mkl_fftw` is set.

**Comment by mtaillefumier** [comment]
> If I build with Makefile
> 
> ```
> cp2k@2023.2%gcc +mpi +openmp +elpa +libint +libxc +spglib smm=libxsmm build_system=makefile ^openmpi ^fftw@3 ^netlib-scalapack ^openblas
> ```
> 
> I hit this linking error due to `libxsmm`:
> 
> ```shell
>      817    /usr/bin/mpif90 -c -O2 -funroll-loops -ftree-vectorize -ffree-form -ffree-line-length-none -ggdb -fopenmp -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libint-2.6.0-gzzjdrt4h4fqli2hi75vs34gsxy6lh32/include -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libint-2.6.0-gzzjdrt4h4fqli2hi75vs34gsxy6lh32/include/
>             libint2 -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/boost-1.84.0-hwwnraam3zmigqp2yqy4dipb7cz7ib4q/include -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libxc-6.2.2-4s2i4pso4wjmk3tfpisxi4hqg2ybdta2/include -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/elpa-2023.05.001-kadh2
>             ux2h6p67dkebz6dwzwaiysvqupi/include/elpa_openmp-2023.05.001/modules -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/elpa-2023.05.001-kadh2ux2h6p67dkebz6dwzwaiysvqupi/include/elpa_openmp-2023.05.001/elpa -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libxsmm-1.17-wzwhf5uriw2g4zopxntmjlbx7wx4pk7n/include -D__F
>             FTW3 -I/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/fftw-3.3.10-rvdcxs5iywzcwzrxzwzamexdke5czoki/include -D__parallel -D__SCALAPACK -D__LIBINT -D__LIBXC -D__ELPA=202305 -D__LIBXSMM -D__SPGLIB -D__COMPILE_ARCH="\"linux-ubuntu22.04-skylake-gcc\"" -D__COMPILE_DATE="\"Fri Jan 19 16:33:07 CET 2024\"" -D__COMPILE_HOST="\"DESKTOP-HSCRDM6\"" -D
>             __COMPILE_REVISION="\"git:65cca6c\"" -D__DATA_DIR="\"/home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/cp2k-2023.2-3br2elq4zz6jwu7apjgupo7vab5tfhmm/share/data\"" -D__SHORT_FILE__="\"distribution_2d_types.F\"" -I'/tmp/lercole/spack-stage/spack-stage-cp2k-2023.2-3br2elq4zz6jwu7apjgupo7vab5tfhmm/spack-src/src/' -I'/tmp/lercole/spack-stage/spack
>             -stage-cp2k-2023.2-3br2elq4zz6jwu7apjgupo7vab5tfhmm/spack-src/obj/linux-ubuntu22.04-skylake-gcc/psmp/exts/dbcsr' distribution_2d_types.F90
>   >> 818    /usr/bin/ld: /home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libxsmm-1.17-wzwhf5uriw2g4zopxntmjlbx7wx4pk7n/lib/libxsmm.so: undefined reference to `dgemm_'
>   >> 819    /usr/bin/ld: /home/lercole/src/spack-dev/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/libxsmm-1.17-wzwhf5uriw2g4zopxntmjlbx7wx4pk7n/lib/libxsmm.so: undefined reference to `sgemm_'
> ```

ubuntu.... What a pain.

**Comment by lorisercole** [comment]
Oh I see. Indeed on AlmaLinux it was working fine...

**Comment by alalazo** [comment]
I guess if there are no new failures, we can consider merging this?

**Comment by mtaillefumier** [comment]
this work on Fedora as well 

`spack install cp2k@2024.1 +libxc +libint +mpi smm=libxsmm ^intel-oneapi-mkl+cluster ^openmpi`

so far I only have had issues with ubuntu. It is a real pain when building containers for instance.

**Review by mtaillefumier — APPROVED** [review]
unless we find something else not working this can be merged. Thanks for the hard work

we should maybe add an exclusion for openblas and mkl-fftw if it is possible. Although i do not understand how this can happen.

**Review by mtaillefumier — APPROVED** [review]

**Comment by mtaillefumier** [comment]
@alalazo something is wrong on fedora with 2023.2. I need to check what's going on. So delay the merge if possible. If  I only have issues with it then merge this (when [lorisercole](https://github.com/lorisercole)) give you the green light as well.

Thanks for this. this recipe needed some cleanup.

**Comment by mtaillefumier** [comment]
oh we have an issue here. 
```
        lapack = spec["lapack"]
        blas = spec["blas"]

        if blas.name in ["intel-mkl", "intel-parallel-studio", "intel-oneapi-mkl"]:
            args += ["-DCP2K_BLAS_VENDOR=MKL"]
            if sys.platform == "darwin":
                args += [
                    self.define("CP2K_BLAS_VENDOR", "CUSTOM"),
                    self.define("CP2K_SCALAPACK_VENDOR", "GENERIC"),
                    self.define(
                        "CP2K_SCALAPACK_LINK_LIBRARIES", spec["scalapack"].libs.joined(";")
                    ),
                ]
            else:
                args += ["-DCP2K_SCALAPACK_VENDOR=MKL"]
        else:
            args.extend(
                [
                    self.define("CP2K_LAPACK_FOUND", True),
                    self.define("CP2K_LAPACK_LINK_LIBRARIES", lapack.libs.joined(";")),
                    self.define("CP2K_BLAS_FOUND", True),
                    self.define("CP2K_BLAS_LINK_LIBRARIES", blas.libs.joined(";")),
                    self.define("CP2K_SCALAPACK_FOUND", True),
                    self.define("CP2K_SCALAPACK_INCLUDE_DIRS", spec["scalapack"].prefix.include),
                    self.define("CP2K_BLAS_VENDOR", "CUSTOM"),
                    self.define("CP2K_SCALAPACK_VENDOR", "GENERIC"),
                    self.define(
                        "CP2K_SCALAPACK_LINK_LIBRARIES", spec["scalapack"].libs.joined(";")
                    ),
                ]
            )

```

this block does not seem to work as I thought it would.
`
spack install cp2k@2024.1 +libxc +libint +mpi smm=libxsmm ^intel-oneapi-mkl+cluster ^openmpi` 

for instance returns mkl for scalapack and openblas on my system for no obvious reasons. so the else statement is taken despite specifying mkl in the spec... Strange

**Comment by mtaillefumier** [comment]
this works 

spack install "cp2k@2024.1 +libxc +libint +mpi smm=libxsmm ^[virtuals=blas,lapack,scalapack] intel-oneapi-mkl +cluster ^openmpi"
spack install "cp2k@2023.2 +libxc +libint +mpi smm=libxsmm ^[virtuals=blas,lapack,scalapack] intel-oneapi-mkl +cluster ^openmpi"

I forgot about the virtual dependency.

**Review by mtaillefumier — APPROVED** [review]
legit

**Comment by alalazo** [comment]
There seem to be a legit failure in pipelines, with `gromacs` not being able to find `cp2k`

**Comment by lorisercole** [comment]
> @alalazo something is wrong on fedora with 2023.2. I need to check what's going on. So delay the merge if possible. If I only have issues with it then merge this (when [lorisercole](https://github.com/lorisercole)) give you the green light as well.
> 
> Thanks for this. this recipe needed some cleanup.

All good for me!

**Comment by alalazo** [comment]
Since this has the :heavy_check_mark: of two maintainers, I'm merging it. Let me know if there is any follow up needed.

**Pull request merged**
